drivers: gpio: add two shared priority symbols for I2C and SPI devices

fabiobaltieri opened the pull request:
There's few SPI and I2C GPIO controllers being added, each one adding a custom initialization priority symbol. Adding two shared priority symbols for I2C and SPI devices respectively to remove some config options.

Those new options are hardcoded with the bus priority+1, took the chance to also changed the default SPI controller priority to `KERNEL_INIT_PRIORITY_DEVICE`, matching how other bus controllers are defined.

Keeping two symbols separate just in case some application needs an usual setup, like bitbanging SPI on an I2C port expander or something like that.

fabiobaltieri committed:
drivers: spi: use KERNEL_INIT_PRIORITY_DEVICE for init priority

Change SPI_INIT_PRIORITY default to use KERNEL_INIT_PRIORITY_DEVICE
instead of a hardcoded value. This brings it inline with other Zephyr
bus configurations, including I2C.

The resulting priority changes from 70 to 50.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
drivers: gpio_s32: use the I2C priority symbol

Use CONFIG_GPIO_INIT_PRIORITY like the other GPIO drivers for the
initialization priority of the gpio_s32 driver rather than hardcoding
CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, which is the default value anyway.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
drivers: gpio: add two shared priority symbols for I2C and SPI devices

Define two common configuration option for I2C and SPI GPIO controllers,
use them to replace device specific priority config entries where
possible.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

galak commented:
We really should try and utilize the devicetree dependency information to handle init order so such changes are not needed.

fabiobaltieri commented:
> We really should try and utilize the devicetree dependency information to handle init order so such changes are not needed.

Yeah that'd be ideal, but in the meantime maybe we can at least not have one option per driver.

cfriedt commented on drivers/gpio/gpio_sn74hc595.c:
```diff
@@ -21,8 +21,8 @@
 
-#if CONFIG_SPI_INIT_PRIORITY >= CONFIG_GPIO_SN74HC595_INIT_PRIORITY
-#error SPI_INIT_PRIORITY must be lower than SN74HC595_INIT_PRIORITY
+#if CONFIG_SPI_INIT_PRIORITY >= CONFIG_GPIO_SPI_INIT_PRIORITY
```
This could be factored-out and put into common code now (sorry if that's already been done, i just didn't see it).

fabiobaltieri replied:
Yeah was thinking about it wondering how to do it, my concern is that if I just dump it in the common gpio.h header it may end up breaking the build if someone is changing the i2c priority without changing the two new symbols even if they don't use any i2c/spi port expander. Alternatively I could put it in its own local header and include it only on drivers that needs it. What do you think?

cfriedt replied:
Probably the latter option. It is tricky though..

fabiobaltieri replied:
I'll give it a try, see how ugly it gets :-)

fabiobaltieri replied:
Added a patch for that, looks a bit hackish but I can't quite think of a clean solution. Could also add a hidden symbol to control that and then flag it in Kconfig but it'd probably be equally messy. Guess eventually we'll get dt based priority anyway and it'll clean all this mess.

cfriedt commented:
> We really should try and utilize the devicetree dependency information to handle init order so such changes are not needed.

Probably also smart to create an optional DT property so that the same types of devices, even on the same bus, can be given explicit init priority.

fabiobaltieri commented:
> > We really should try and utilize the devicetree dependency information to handle init order so such changes are not needed.
> 
> Probably also smart to create an optional DT property so that the same types of devices, even on the same bus, can be given explicit init priority.

If we really want to go that way I think should add a priority label in the base property and use it automatically from the `DEVICE_DT_INST_DEFINE` macro. Maybe it could be auto-populated by the EDT libraries and pave the way to remove all of this priority stuff. Regardless, feels like a major change that deserves its own design and discussion, not too keen to start adding custom properties here unless you feel strongly about it.

fabiobaltieri committed:
drivers: gpio: add spi and i2c priority build check

Add a couple of build time check to ensure that the priority of i2c or
spi gpio devices is set correctly. To avoid triggering this in systems
where no i2c/spi gpio devices are used, only add the check in
potentially affected drivers.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
- Added common priority checking headers

henrikbrixandersen reviewed (changes requested):
Given that we have just agreed to delay merging #51217 with the expectation that the SYS_INIT/device initialization order methods will change in the somewhat near future I do not think we should introduce this breaking Kconfig change at this point either.

henrikbrixandersen commented on drivers/spi/Kconfig:
```diff
@@ -37,7 +37,7 @@ config SPI_EXTENDED_MODES
 config SPI_INIT_PRIORITY
 	int "Init priority"
-	default 70
+	default KERNEL_INIT_PRIORITY_DEVICE
```
This small change (70 to 50) has the potential of breaking a lot of implicit dependencies on initialization priorities.

fabiobaltieri replied:
Yeah, sorry did not mean to try sneak the change in. I'm not too attached to this change and happy to drop that patch form the stack and adjust the GPIO_SPI default accordingly if there's consensus about that - but - I can't figure a case to not set SPI devices at the same level as I2C, all drivers are either memory mapped or needs GPIO and that's it. Sure there may be some edge case like soft core or something like that, but I imagine that'd be using explicit priorities anyway, and since the change make the priority higher, I feel like it should not break much, and it'd be nice to have similar setup for bus drivers with similar purposes.

@tbursztyka fyi

tbursztyka replied:
Actually, this is an old value set at a time where indeed gpio and spi could be initiated in the wrong order. Afaik, it is not the case anymore since GPIO gets initialized at KERNEL_INIT_PRIORITY_DEFAULT (which was not the case when this above value was set).

Should be fine I think

henrikbrixandersen commented:
> We really should try and utilize the devicetree dependency information to handle init order so such changes are not needed.

Agreed. We really need to stop making minor changes to the existing priorities and levels like this without addressing the main issue. The dependency issues introduced are very difficult to spot as they only show up at run-time.

fabiobaltieri commented:
> > We really should try and utilize the devicetree dependency information to handle init order so such changes are not needed.
> 
> Agreed. We really need to stop making minor changes to the existing priorities and levels like this without addressing the main issue. The dependency issues introduced are very difficult to spot as they only show up at run-time.

Well this adds build time priority check for all GPIO SPI/I2C drivers, so I'd say it's a step forward on that front. :-)

pdgendt commented on drivers/gpio/gpio_i2c_priority_check.h:
```diff
@@ -0,0 +1,10 @@
+#include <zephyr/toolchain/common.h>
+
+BUILD_ASSERT(CONFIG_GPIO_I2C_INIT_PRIORITY > CONFIG_I2C_INIT_PRIORITY,
+	     "GPIO_I2C_INIT_PRIORITY has be lower (higher value) than I2C_PRIORITY");
```
```suggestion
	     "GPIO_I2C_INIT_PRIORITY has to be lower (higher value) than I2C_PRIORITY");
```

pdgendt commented on drivers/gpio/gpio_spi_priority_check.h:
```diff
@@ -0,0 +1,10 @@
+#include <zephyr/toolchain/common.h>
+
+BUILD_ASSERT(CONFIG_GPIO_SPI_INIT_PRIORITY > CONFIG_SPI_INIT_PRIORITY,
+	     "GPIO_SPI_INIT_PRIORITY has be lower (higher value) than SPI_PRIORITY");
```
```suggestion
	     "GPIO_SPI_INIT_PRIORITY has to be lower (higher value) than SPI_PRIORITY");
```

fabiobaltieri commented:
Hey I updated this to drop even more symbols, what do you think? I don't mind dropping it if there's consensus that it's not worth it but it'd also be nice to slow down the proliferation of custom priority symbols until we find a better solution.

fabiobaltieri commented:
Rebased once again, hey @henrikbrixandersen, should I drop this forever? :-) Really I'm not that attached to it, if that's a definitive NACK I don't mind.

henrikbrixandersen commented:
> Rebased once again, hey @henrikbrixandersen, should I drop this forever? :-) Really I'm not that attached to it, if that's a definitive NACK I don't mind.

While I appreciate your efforts in attempting to address this, I still do not think breaking existing configurations with an interim solution is the right way to go. Perhaps bring it up at the architecture WG meeting?

fabiobaltieri commented:
> > Rebased once again, hey @henrikbrixandersen, should I drop this forever? :-) Really I'm not that attached to it, if that's a definitive NACK I don't mind.
> 
> While I appreciate your efforts in attempting to address this, I still do not think breaking existing configurations with an interim solution is the right way to go. Perhaps bring it up at the architecture WG meeting?

For the SPI priority change you mean? Are you confident it's going to break something? I don't think it would.

fabiobaltieri commented:
From offline chat, let's not bother, we can always revisit at a later stage. :-)
